### ISSUE TITLE
Data Seeding: Add support for navigations

### ISSUE THREAD

**Issue by AndriySvyryd** [aspnet/EntityFrameworkCore#10000]
This depends on #9914

**Comment by smitpatel** [comment]
5 figures :trollface:

**Comment by AndriySvyryd** [comment]
This was completely not on purpose. It just happens to be that I had to file 5 issues today.

**Comment by Eilon** [comment]
I think this is the first repo in the 'aspnet' org with this number of issues. Do we celebrate?

**Comment by iamkarlson** [comment]
Is there really no workaround till you're busy implementing?

**Comment by AndriySvyryd** [comment]
@iamkarlson The workaround is to specify the foreign key values. See https://docs.microsoft.com/en-us/ef/core/modeling/data-seeding

**Comment by iamkarlson** [comment]
@AndriySvyryd Thanks, actually I did how you said but the problem was that even an empty list raises such error. I had to manually set it to null after my lazy loading.

**Comment by AndriySvyryd** [comment]
@iamkarlson that sounds like a different issue. Could you file it separately and include a small repro?

**Comment by iamkarlson** [comment]
Do you mind if I attach #12688 to that, just to make sure it won't be missed when you complete this task?

**Comment by AndriySvyryd** [comment]
I still don't know what is the scenario in #12688

**Comment by tohidazizi** [comment]
While both #12004 and #12688 are closed, isn't it the right time to work on this issue?

**Comment by iamkarlson** [comment]
Guys, any ideas about a timeline for this feature?

**Comment by AndriySvyryd** [comment]
It's in the backlog, so it probably will be done after 2019. Remember to vote :+1: for features you'd like us to prioritize.
@iamkarlson If you just want it to not throw for empty collections please file a new issue.

**Comment by weitzhandler** [comment]
I find this issue very annoying, I have an owned entity that I want to set its property at the parent to auto-initialize if null:

```c#
class Contact
{
  public int Id { get; set; }

  Address _Address;
  public Address Address
  {
    get => _Address ??= new Address { /*tried this too*/ ContactId = Id };
    set => _Address = value;
  }
}
```

When adding the parent as `HasData`  I get the following exception:

>System.InvalidOperationException: 'The seed entity for entity type 'Contact' with the key value 'Id:3' cannot be added because it has the navigation 'Address' set. To seed relationships you need to add the related entity seed to 'Address' and specify the foreign key values {'ContactId'}.'

Would love to see this functionality enabled, so we can use auto-initialized properties in entities.

**Comment by weitzhandler** [comment]
Solution was adding the parent entity as anonymous class:

```c#
modelBuilder.Entity<Contact>().HasData(new { ... });
```

**Comment by Unders0n** [comment]
is there any update on this feature? Seems for me like there's not much sense in Seeding mechanism if we need to use anonymous classes for all owned properties. It's much easier to use oldfashioned manual seeding on startup, isn't it?

**Comment by ajcvickers** [comment]
@Unders0n While data in the model can be useful, I would agree that manual seeding on startup is often a better approach.

**Comment by ggonzalez94** [comment]
Any update on when you plan to add this? For configuration data that should be only changed with the source code having data seeding is very useful, but having to depend on anonymous types is error prone and less readable

**Comment by ajcvickers** [comment]
@ggonzalez94 This issue is in the Backlog milestone. This means that it is not planned for the next release (EF Core 5.0). We will [re-assess the backlog](https://docs.microsoft.com/ef/core/what-is-new/release-planning) following the this release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by pantonis** [comment]
Any workaround for optional relationships?

**Comment by rcollette** [comment]
Seeding Owned Entities gets even uglier if you have to implement the hack to allow composite indexes comprised of properties from both the owning and owned entity types.  Now I have to seed the shadow property(ies) that were created for the indexing hack.

```c#
            OwnedNavigationBuilder<SoftwareRelease, SemanticVersion> softwareReleaseNavigationBuilder = null!;
            modelBuilder.Entity<SoftwareRelease>()
                .OwnsOne(p => p.SemanticVersion,
                    oe =>
                    {
                        // Workaround for the fact that owned entities cannot currently be used in a composite key with properties of the owning entity.
                        // Shadow the owning entity property into the owned entity.
                        // https://github.com/dotnet/efcore/issues/11336
                        oe.Property<string>("ShadowProductDeliveryGroupCode")
                            .HasColumnName("product_delivery_group_code");
                        oe.HasIndex("ShadowProductDeliveryGroupCode", "Major", "Minor", "Patch");
                        softwareReleaseNavigationBuilder = oe;
                    });
```
OwnedNavigationBuilder later used to seed data.

```c#
        private static void SeedSoftwareReleaseVersions(OwnedNavigationBuilder<SoftwareRelease,SemanticVersion> builder)
        {
            builder
                .HasData(
                    // When seeding Owned entities, we have to use an anonymous type because the primary key field is a shadow field
                    CreateSoftwareReleaseVersion(1, "6.0.17", "code1"),
                    CreateSoftwareReleaseVersion(2, "6.0.18", "code1"),
                    CreateSoftwareReleaseVersion(3, "9.9.5", "code2"),
                    CreateSoftwareReleaseVersion(4, "9.9.6", "code2"),
                    CreateSoftwareReleaseVersion(5, "9.9.7", "code2"),
                    CreateSoftwareReleaseVersion(6, "9.9.8", "code2")
                );
        }

        private static dynamic CreateSoftwareReleaseVersion(long softwareReleaseId, string versionDisplay,string shadowProductDeliveryGroupCode)
        {
            SemanticVersion version = new(versionDisplay);
            return new
            {
                version.Major,
                version.Minor,
                version.Patch,
                SoftwareReleaseId = softwareReleaseId,
                version.Display,
                ShadowProductDeliveryGroupCode = shadowProductDeliveryGroupCode
            };
        }
```

References:
#11336
#12004
https://docs.microsoft.com/en-us/ef/core/modeling/data-seeding
https://docs.microsoft.com/en-us/archive/msdn-magazine/2018/august/data-points-deep-dive-into-ef-core-hasdata-seeding

**Comment by slawton-concurrency** [comment]
How is there still no update for this 5 years later?

**Comment by hkusulja** [comment]
So many github issues linked to this, and depend on this.
It passed so many years.
It is still tagged as consider for current release.

Any update on this?
